src/cript/api/api.py

```diff
@@ -513,6 +520,11 @@ def save(self, project: Project) -> None:
         None
             Just sends a `POST` or `Patch` request to the API
         """
+
+        # Ensure that all file nodes have uploaded there payload before actual save.
+        for file_node in project.find_children({"node": ["File"]}):
```

Comment:
this is an O(n) operation and `find_children` is an o(n^2) operation, this is total of an o(n^3) operation. This gets me worried that it could be slow very quickly if someone is using our script with something bigger.

How essential would you say it is for upload to happen on save instead of when the file node is created?

Reply:
No, it is not.
First, what do you mean by `n`?

I made a detailed runtime analysis for `find_children` and the runtime is `O(n*(k+m*h)`, with `n` being the number of nodes, `k` being the number of children of the node (assume max. if worst case), `m` being the number of nodes in the searched for attribute, and `h` being the depth of the search dictionary.

I added comments into the code, show this.

Back to this usage: The depth of the search dictionary is 1, we dont' know number of children and length of the attribute lists, but they aren't long. Guaranteed n > m, n > k.

So, the actual `find_children` call is closer to being a linear operation in the number of nodes then to be O(n**2).

And the second loop, is linear in the number of actual file nodes.
(Which you have to be linear in, since we have to upload them all).

So, this is not an issue.

Reply:
We could consider rewriting `find_children` as a generator.
In that case, we can eliminate the second loops cost at all.

Reply:
ok, this is implementation details of find_children.
We can later reimplement it as we find better solutions.

(The best that comes to mind, is that we have a variety of caches (uuid hash based), that let us look up stuff.
For example: we have a cache of all "file" nodes, and we keep a look-up map for all nodes in a project.
But both of those ideas require additional code, making the SDK less lightweight.)

I am open to add something like this, later so far we have pushed stuff like this back.